cli/listener/src/main.rs

```diff
@@ -138,24 +138,51 @@ pub fn get_apy_for_period(
             slot: row.get(2)?,
             epoch: row.get(3)?,
             pool: row.get(4)?,
-            token_a: row.get(5)?,
-            token_b: row.get(6)?,
+            price_lamports_numerator: row.get(5)?,
+            price_lamports_denominator: row.get(6)?,
         })
     };
 
-    // Get first logged minimal logged data based on timestamp that is greater than `from_time`.
-    let mut stmt_first = conn.prepare(
-        "SELECT *, MIN(timestamp) FROM exchange_rate WHERE pool = :pool AND timestamp > :t",
-    )?;
-    let mut row_iter = stmt_first.query_map([opts.pool.clone(), from_time.to_string()], row_map)?;
-    let first = row_iter.next();
+    let (first, last) = {
+        // Get first logged minimal logged data based on timestamp that is greater than `from_time`.
+        let stmt_first = &mut tx.prepare(
+            "WITH prices_epoch AS (
+                SELECT *
+                FROM exchange_rate
+                WHERE epoch = (SELECT MIN(epoch) from exchange_rate where pool = :pool AND timestamp > :t)
+              )
+              SELECT
+                *
+              FROM
+                prices_epoch
+              ORDER BY
+                timestamp ASC
+            ",
+        )?;
+        // Get first logged maximal logged data based on timestamp that is smaller than `to_time`.
+        let stmt_last =
+            &mut tx.prepare("WITH prices_epoch AS (
+                SELECT *
+                FROM exchange_rate
+                WHERE epoch = (SELECT MAX(epoch) from exchange_rate where pool = :pool AND timestamp < :t)
+              )
+              SELECT
+                *
+              FROM
+                prices_epoch
+              ORDER BY
+                timestamp ASC
+            ")?;
+        let mut row_iter =
+            stmt_first.query_map([opts.pool.clone(), from_time.to_string()], row_map)?;
+        let first = row_iter.next();
 
-    // Get first logged maximal logged data based on timestamp that is smaller than `to_time`.
-    let mut stmt_last = conn.prepare(
-        "SELECT *, MAX(timestamp) FROM exchange_rate WHERE pool = :pool AND timestamp < :t",
-    )?;
-    let mut row_iter = stmt_last.query_map([opts.pool.clone(), to_time.to_string()], row_map)?;
-    let last = row_iter.next();
+        let mut row_iter =
+            stmt_last.query_map([opts.pool.clone(), to_time.to_string()], row_map)?;
+        let last = row_iter.next();
+        (first, last)
+    };
+    tx.commit()?;
```

Comment:
Why the commit here? This is a read-only function right? It shouldn’t modify the transaction. I’m not sure how the types work, but it would be nice if we can take the `rusqlite::Transaction` as a non-mutable reference to enforce that.